File: src/game/systems/attack.lua
local Aspect = require 'ecs.aspect'
local System = require 'ecs.system'
local Ability = require 'game.components.ability'
local Attack = require 'game.components.attack'
local Fixture = require 'game.components.fixture'
local Movement = require 'game.components.movement'
local Position = require 'game.components.position'

local aspect = Aspect.new({Ability, Attack, Fixture, Movement, Position})
local AttackSystem = System:new('aggression', aspect)

function AttackSystem:onRemove(entity)
  entity:as(Ability):reset()
end

function AttackSystem:shoot(dt, entity)
  local ability = entity:as(Ability)
  local attack = entity:as(Attack)
  local movement = entity:as(Movement)
  local x1, y2 = entity:as(Position):coords()
  local x2, y2 = attack.target:as(Position):coords()

  if x2 <= x1 then
    movement.direction = 'left'
  else
    movement.direction = 'right'
  end

  ability:setActivated('shoot', true)
end

function AttackSystem:slash(dt, entity)
  local ability = entity:as(Ability)
  local attack = entity:as(Attack)
  local fixture = entity:as(Fixture)
  local position = entity:as(Position)
  local position2 = attack.target:as(Position)
  local distance = position2.x - position.x
  local body = fixture.fixture:getBody()
  local velocityX, velocityY = body:getLinearVelocity()
  local newVelocityX = math.sign(distance) * 100

  if math.abs(distance) < 50 then
    newVelocityX = 0
    ability:setActivated('slash', true)
  end

  body:setLinearVelocity(newVelocityX, velocityY)
end

function AttackSystem:update(dt)
  for _, entity in pairs(self.entities) do
    local abilities = entity:as(Ability).abilities

    if abilities.slash.enabled then
      self:slash(dt, entity)
    elseif abilities.shoot.enabled then
      self:shoot(dt, entity)
    end
  end
end

return AttackSystem

File: client2/src/game/systems/Attack.ts
import { Aspect } from '../../ecs/Aspect'
import { Entity } from '../../ecs/Entity'
import { System } from '../../ecs/System'
import { Abilities, AbilityType } from '../components/Abilities'
import { Attack } from '../components/Attack'
import { GameObject } from '../components/GameObject'
import { Movement } from '../components/Movement'
import { Position } from '../components/Position'
import { SystemFlag } from '../flags'
import * as math from '../utils/math'

const getDistanceToTarget = (target: Entity, entity: Entity): number => {
  const targetPosition = target.as(Position)
  const entityPosition = entity.as(Position)

  if (targetPosition && entityPosition) {
    return targetPosition.x - entityPosition.x
  }

  return 0
}

export class AttackSystem extends System {
  static _id = 'Attack'
  _id = AttackSystem._id

  static _flag = SystemFlag.Attack
  _flag = AttackSystem._flag

  static _aspect = new Aspect([Attack, GameObject, Movement, Position])
  _aspect = AttackSystem._aspect

  update = (dt: number) => {
    this.entities.forEach(entity => {
      const attack = entity.as(Attack)
      const gameObject = entity.as(GameObject)
      const movement = entity.as(Movement)

      if (!attack || !gameObject || !movement) {
        return
      }

      const dx = getDistanceToTarget(attack.target, entity)

      if (Math.abs(dx) < attack.followDistance) {
        // in range, activate abilities
        const abilities = entity.as(Abilities)

        if (abilities) {
          for (let key in abilities.abilities) {
            const ability = abilities.abilities[key as AbilityType]

            if (ability) {
              ability.activated = true
            }
          }
        }
      } else {
        // not in range, follow target
        const body = gameObject.fixture.getBody()

        let [velocityX, velocityY] = body.getLinearVelocity()

        velocityX = 0

        if (Math.abs(dx) > attack.followDistance) {
          velocityX = math.sign(dx) * attack.followVelocity
        }

        body.setLinearVelocity(velocityX, velocityY)
      }

      // make sure attacker is facing the target
      if (dx < 0) {
        movement.direction = 'left'
      } else if (dx > 0) {
        movement.direction = 'right'
      }
    })
  }

  onRemove = (entity: Entity) => {
    const abilities = entity.as(Abilities)

    abilities && abilities.reset()
  }
}
